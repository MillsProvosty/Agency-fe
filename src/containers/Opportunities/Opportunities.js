import React, { useState } from "react";
import { connect } from "react-redux";
import Modal from "react-modal";
import CreateOppModal from "../../components/CreateOppModal/CreateOppModal";
import { editOpp } from "../../actions";
import { deleteAnOpportunity } from "../../util/apiCalls";
import styled from "styled-components";
import { Link } from "react-router-dom";

const OpportunityCard = styled.section`
  width: 90%;
  margin: auto;
  display: flex;
  margin-top: 0.5em;
  padding-bottom: 0.5em;
  align-items: center;
  height: 15vh;
  background-color: white;
  border: 1px solid #37474e;
  -webkit-box-shadow: 9px 10px 5px -10px rgba(55, 71, 78, 1);
  -moz-box-shadow: 9px 10px 5px -10px rgba(55, 71, 78, 1);
  box-shadow: 9px 10px 5px -10px rgba(55, 71, 78, 1);
  border: 1px solid #37474e @media screen and (max-width: 375px) {
    border-right: none;
  }
  @media screen and (display-mode: standalone) {
    border-right: none;
  }
`;

const OpportunitySection = styled.section`
  width: 100%;
  background-color: aliceblue;
  background-attachment: fixed;
`;

const PTag = styled.p`
  font-size: 1em};
  margin: 0px;
`;

const Header = styled.h1`
  color: #37474e;
  font-size: 2em;
  margin: 0px;
  @media screen and (max-width: 375px) {
    margin: 0px 0px 30px -80px;
  }
`;

const ModalStyle = styled.section`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`;

const Button = styled.button`
  color: white;
  background-color: #37474e;
  border-radius: 5px;
  font-size: 1em;
  width: 175px;
  font-family: "Quicksand", sans-serif;
  border: 2px solid white;
  margin-right: ${props => (props.volunteer ? "20px" : "0px")};
  margin-left: ${props => (props.client ? "20px" : "0px")} @media screen and
    (max-width: 375px) {
    margin: ${props =>
      props.client ? "20px 0px 0px 0px" : "50px 20px 0px 0px"};
  }

  :hover {
    border: 2px solid #37474e;
    color: #37474e;
    background-color: white;
  }
`;

const CardSection = styled.section`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: start;
  height: 100%;
  width: ${props => (props.buttons ? "20%" : "35%")};
`;

const Container = styled.section`
  width: 100%;
`;

const Bold = styled.span`
  font-weight: bold;
`;

export const Opportunities = props => {
  const deleteOpportunity = async (userId, oppId) => {
    await deleteAnOpportunity(userId, oppId);
  };

  const displayOpp = () => {
    return props.opportunities.map(opportunity => {
      return (
        <OpportunityCard key={opportunity.id}>
          <CardSection>
            <Header>{opportunity.title}</Header>
            <PTag description>{opportunity.description}</PTag>
          </CardSection>
          <CardSection>
            <PTag>
              <Bold>Type: </Bold>
              {opportunity.type}
            </PTag>
            <PTag>
              <Bold>Location:</Bold> {opportunity.location}
            </PTag>
            <PTag>
              <Bold>Estimated Time:</Bold>
              {opportunity.estimated_time}
            </PTag>
          </CardSection>
          {props.role === "volunteer" && (
            <CardSection buttons>
              <Button
                onClick={() =>
                  deleteAnOpportunity(props.user.id, opportunity.id)
                }
              >
                Select opportunity; doesnt work
              </Button>
              <Button
                onClick={() =>
                  deleteAnOpportunity(props.user.id, opportunity.id)
                }
              >
                Find Out About The Client; doesnt work
              </Button>
            </CardSection>
          )}
          {props.role === "client" && (
            <CardSection buttons>
              <Button
                onClick={() => deleteOpportunity(props.user.id, opportunity.id)}
              >
                DELETE works and throws error
              </Button>
              <Button onClick={props.editOpp}>This Edit Doesnt Work</Button>
            </CardSection>
          )}
        </OpportunityCard>
      );
    });
  };

  const [createModal, showCreateModal] = React.useState(false);

  return (
    <OpportunitySection>
      <ModalStyle>
        <Modal isOpen={createModal} className="modal">
          <CreateOppModal />
        </Modal>
      </ModalStyle>
      <Link to="/profile">
        <Button>Return to Profile</Button>
        <Button>Edit my Profile; doesnt work</Button>
      </Link>
      {props.role === "client" && (
        <Button id='showModal'onClick={() => showCreateModal(true)}>
          Create an opportunity
        </Button>
      )}
      {props.role === "volunteer" && <Button>Search For Opportunities</Button>}
      <Container>{displayOpp()}</Container>
    </OpportunitySection>
  );
};

export const mapStateToProps = state => ({
  opportunities: state.opportunities,
  user: state.user
});

export const mapDispatchToProps = dispatch => ({
  editOpp: opp => dispatch(editOpp(opp))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Opportunities);
