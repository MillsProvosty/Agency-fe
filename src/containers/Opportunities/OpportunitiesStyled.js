import styled from "styled-components";


export const OpportunityCard = styled.section`
  width: 90%;
  margin: auto;
  display: flex;
  margin-top: 0.5em;
  padding-bottom: 0.5em;
  margin-bottom: 1em;
  align-items: center;
  height: 15vh;
  background-color: aliceblue;
  border: 1px solid #37474e;
  -webkit-box-shadow: 9px 10px 5px -10px rgba(55, 71, 78, 1);
  -moz-box-shadow: 9px 10px 5px -10px rgba(55, 71, 78, 1);
  box-shadow: 9px 10px 5px -10px rgba(55, 71, 78, 1);
  border: 1px solid #37474e @media screen and (max-width: 375px) {
    border-right: none;
  }
  @media screen and (display-mode: standalone) {
    border-right: none;
  }
`;

export const OpportunitySection = styled.section`
  width: 100%;
  background-color: white;
  background-attachment: fixed;
`;

export const PTag = styled.p`
  font-size: 1em};
  margin: 0px;
`;

export const Header = styled.h1`
  color: #37474e;
  font-size: 2em;
  margin: 0px;
  @media screen and (max-width: 375px) {
    margin: 0px 0px 30px -80px;
  }
`;

export const ModalStyle = styled.section`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`;

export const Button = styled.button`
  color: white;
  background-color: #37474e;
  border-radius: 5px;
  font-size: 1em;
  width: 175px;
  font-family: "Quicksand", sans-serif;
  border: 2px solid white;
  margin-right: ${props => (props.volunteer ? "20px" : "0px")};
  margin-left: ${props => (props.client ? "20px" : "0px")} @media screen and
    (max-width: 375px) {
    margin: ${props =>
      props.client ? "20px 0px 0px 0px" : "50px 20px 0px 0px"};
  }

  :hover {
    border: 2px solid #37474e;
    color: #37474e;
    background-color: white;
  }
`;

export const CardSection = styled.section`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: start;
  height: 100%;
  width: ${props => (props.buttons ? "20%" : "35%")};
`;

export const Container = styled.section`
  width: 100%;
`;

export const Bold = styled.span`
  font-weight: bold;
  color: ${props => (props.fulfill ? "red" : "#37474e")}
`;
