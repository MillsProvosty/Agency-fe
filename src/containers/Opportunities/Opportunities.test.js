import React from 'react';
import { shallow } from 'enzyme';
import { Opportunities, mapStateToProps } from './Opportunities'

describe('Opportunities', () => {
  let wrapper;
  
  const mockOpp = [
    {
      "description": "I cannot rake my yard- I need someone to come in and get the front and back yard.",
      "estimated_time": "1 hr",
      "id": 2,
      "location": "2848 Roslyn St., Denver CO 80238",
      "title": "Rake Leaves",
      "type": "Physical Labor",
      "user_id": 1
    }
  ];
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, 'useState')
  useStateSpy.mockImplementation((init) => [init, setState]);

  beforeEach(() => {
    wrapper = shallow(
      <Opportunities 
        opportunities={mockOpp}
        role="client"
      />
    )
  })

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  });

  it('should render a different button if role is different', () => {
   const wrapper = shallow(
      <Opportunities 
        opportunities={mockOpp}
        role="volunteer"
      />
   )
   expect(wrapper).toMatchSnapshot()
  })

 it('should call showCreateModal when clicked', () => {
   wrapper.find('#showModal').props().onClick()
   expect(setState).toHaveBeenCalled()
 }) 

  describe('mapStateToProps', () => {
    it('should return an array of opportunities', () => {
      let mockState = {
        user: {
          id: 1,
          name: 'something'
        },
        opportunities: mockOpp,
        role: 'client'
      }
      let expected = mockOpp
  
      let mappedProps = mapStateToProps(mockState);
  
      expect(mappedProps.opportunities).toEqual(expected)
    })

    it('should return a user', () => {
      let mockState = {
        user: {
          id: 1,
          name: 'something'
        },
        opportunities: mockOpp,
        role: 'client'
      }
      let expected = mockState.user
      let mappedProps = mapStateToProps(mockState);
  
      expect(mappedProps.user).toEqual(expected)
    })
  })
})