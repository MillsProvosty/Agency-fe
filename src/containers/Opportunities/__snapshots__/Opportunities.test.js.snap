// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Opportunities should match snapshot 1`] = `
<styled.section>
  <styled.section>
    <Modal
      ariaHideApp={true}
      bodyOpenClassName="ReactModal__Body--open"
      className="modal"
      closeTimeoutMS={0}
      isOpen={false}
      parentSelector={[Function]}
      portalClassName="ReactModalPortal"
      role="dialog"
      shouldCloseOnEsc={true}
      shouldCloseOnOverlayClick={true}
      shouldFocusAfterRender={true}
      shouldReturnFocusAfterClose={true}
    >
      <Connect(CreateOppModal) />
    </Modal>
  </styled.section>
  <Link
    to="/profile"
  >
    <styled.button>
      Return to Profile
    </styled.button>
    <styled.button>
      Edit my Profile; doesnt work
    </styled.button>
  </Link>
  <styled.button
    id="showModal"
    onClick={[Function]}
  >
    Create an opportunity
  </styled.button>
  <styled.section>
    <styled.section
      key="2"
    >
      <styled.section>
        <styled.h1>
          Rake Leaves
        </styled.h1>
        <styled.p
          description={true}
        >
          I cannot rake my yard- I need someone to come in and get the front and back yard.
        </styled.p>
      </styled.section>
      <styled.section>
        <styled.p>
          <styled.span>
            Type: 
          </styled.span>
          Physical Labor
        </styled.p>
        <styled.p>
          <styled.span>
            Location:
          </styled.span>
           
          2848 Roslyn St., Denver CO 80238
        </styled.p>
        <styled.p>
          <styled.span>
            Estimated Time:
          </styled.span>
          1 hr
        </styled.p>
      </styled.section>
      <styled.section
        buttons={true}
      >
        <styled.button
          onClick={[Function]}
        >
          DELETE works and throws error
        </styled.button>
        <styled.button>
          This Edit Doesnt Work
        </styled.button>
      </styled.section>
    </styled.section>
  </styled.section>
</styled.section>
`;

exports[`Opportunities should render a different button if role is different 1`] = `
<styled.section>
  <styled.section>
    <Modal
      ariaHideApp={true}
      bodyOpenClassName="ReactModal__Body--open"
      className="modal"
      closeTimeoutMS={0}
      isOpen={false}
      parentSelector={[Function]}
      portalClassName="ReactModalPortal"
      role="dialog"
      shouldCloseOnEsc={true}
      shouldCloseOnOverlayClick={true}
      shouldFocusAfterRender={true}
      shouldReturnFocusAfterClose={true}
    >
      <Connect(CreateOppModal) />
    </Modal>
  </styled.section>
  <Link
    to="/profile"
  >
    <styled.button>
      Return to Profile
    </styled.button>
    <styled.button>
      Edit my Profile; doesnt work
    </styled.button>
  </Link>
  <styled.button>
    Search For Opportunities
  </styled.button>
  <styled.section>
    <styled.section
      key="2"
    >
      <styled.section>
        <styled.h1>
          Rake Leaves
        </styled.h1>
        <styled.p
          description={true}
        >
          I cannot rake my yard- I need someone to come in and get the front and back yard.
        </styled.p>
      </styled.section>
      <styled.section>
        <styled.p>
          <styled.span>
            Type: 
          </styled.span>
          Physical Labor
        </styled.p>
        <styled.p>
          <styled.span>
            Location:
          </styled.span>
           
          2848 Roslyn St., Denver CO 80238
        </styled.p>
        <styled.p>
          <styled.span>
            Estimated Time:
          </styled.span>
          1 hr
        </styled.p>
      </styled.section>
      <styled.section
        buttons={true}
      >
        <styled.button
          onClick={[Function]}
        >
          Select opportunity; doesnt work
        </styled.button>
        <styled.button
          onClick={[Function]}
        >
          Find Out About The Client; doesnt work
        </styled.button>
      </styled.section>
    </styled.section>
  </styled.section>
</styled.section>
`;
